nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/CompoundDocumentServiceImpl.java

```diff
@@ -0,0 +1,108 @@
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+import org.apache.commons.io.FilenameUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.common.utils.Path;
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.impl.blob.ZipEntryBlob;
+import org.nuxeo.ecm.platform.filemanager.api.FileImporterContext;
+import org.nuxeo.ecm.platform.filemanager.api.FileManager;
+import org.nuxeo.ecm.platform.filemanager.service.FileManagerService;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+/** @since 2021.0 */
+public class CompoundDocumentServiceImpl extends DefaultComponent implements CompoundDocumentService {
+
+    private static final Logger log = LogManager.getLogger(CompoundDocumentServiceImpl.class);
+
+    @Override
+    public DocumentModel createCompoundDocument(CoreSession session, String parent, Blob archive) {
+        String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
+        DocumentModel compoundDoc = session.createDocument(
+                session.createDocumentModel(parent, compoundDocName, COMPOUND_DOCTYPE));
+        createCompoundDocument(compoundDoc, archive);
+        return compoundDoc;
+    }
+
+    protected boolean isAllowedEntry(ZipEntry entry) {
+        CompoundDocumentManager manager = Framework.getService(CompoundDocumentManager.class);
+        Path path = new Path(entry.getName());
+        // TODO Dynamically getIgnoredFiles depending on the compound type.
+        IgnoredFilesDescriptor ignoredFiles = manager.getIgnoredFiles(DEFAULT_IGNORED_FILES);
+        for (String segment : path.segments()) {
+            if (ignoredFiles.getIgnoredFileDescriptors().stream().anyMatch(i -> i.ignore(segment))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    protected void createCompoundDocument(DocumentModel compoundDoc, Blob archive) {
+        try (ZipFile zip = new ZipFile(archive.getFile())) {
+            FileManagerService fileManager = (FileManagerService) Framework.getService(FileManager.class);
+            zip.stream().filter(this::isAllowedEntry).sorted(Comparator.comparing(ZipEntry::getName)).forEach(entry -> {
+                Path entryPath = new Path(entry.getName());
+                String parentPath = compoundDoc.getPathAsString() + separator
+                        + entryPath.removeLastSegments(1).toString();
+                if (entry.isDirectory()) {
+                    CoreSession session = compoundDoc.getCoreSession();
+                    session.createDocument(
+                            session.createDocumentModel(parentPath, entryPath.lastSegment(), COMPOUND_FOLDER_DOCTYPE))
```

Comment:
it will improve readability to have two lines here, one for `createDocumentModel` and one for `createDocument`.